lcls-twincat-motion/Library/POUs/Motion/FB_MotionRequest.TcPOU

```diff
@@ -0,0 +1,178 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4022.18">
+  <POU Name="FB_MotionRequest" Id="{f3ce72e2-b968-4260-9a64-8e0267e89e51}" SpecialFunc="None">
+    <Declaration><![CDATA[FUNCTION_BLOCK FB_MotionRequest
+(*
+	Request a move from an axis controlled via EPICS using FB_MotionStage
+	This exists to manage situations where different bits of code may need to move the same motor.
+	With just the DUT_MotionStage/FB_MotionStage setup it is possible for two function blocks to
+	fight with and interfere with each other and with the EPICS commands.
+*)
+VAR_IN_OUT
+	// Motor to move
+    stMotionStage: DUT_MotionStage;
+END_VAR
+VAR_INPUT
+	// Start move on rising edge, stop move on falling edge
+    bExecute: BOOL;
+	// Reset errors on rising edge
+    bReset: BOOL;
+	// Define behavior for when the motor is already moving
+    enumMotionRequest: ENUM_MotionRequest := ENUM_MotionRequest.WAIT;
+	// Goal position
+    fPos: LREAL;
+	// Move velocity
+    fVel: LREAL;
+	// Optional acceleration
+    fAcc: LREAL;
+	// Optional deceleration
+    fDec: LREAL;
+END_VAR
+VAR_OUTPUT
+	// True if in error state
+    bError: BOOL;
+	// Error code
+    nErrorId: UDINT;
+    // What the error code means
+    sErrorMessage: STRING;
+	// If TRUE, we are moving the motor
+    bBusy: BOOL;
+	// If TRUE, we are not moving the motor and our most recent move was successful
+    bDone: BOOL;
+END_VAR
+VAR
+    rtExec: R_TRIG;
+    ftExec: F_TRIG;
+    rtReset: R_TRIG;
+    ftBusy: F_TRIG;
+    nState: UINT := 0;
+    bMyMove: BOOL;
+    bCausedError: BOOL;
+END_VAR
+// Define local constants for our state machine states
+VAR CONSTANT
+    INIT: UINT := 0;
+    WAIT_EXEC: UINT := 1;
+    PICK_REQUEST: UINT := 2;
+    WAIT_OTHER_MOVE: UINT := 3;
+    STOP_OTHER_MOVE: UINT := 4;
+    START_MOVE: UINT := 5;
+    WAIT_MY_MOVE: UINT := 6;
+    STOP_MY_MOVE: UINT := 7;
+    DONE_MOVING: UINT := 8;
+    ERROR: UINT := 9;
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[rtExec(CLK:=bExecute);
+ftExec(CLK:=bExecute);
+rtReset(CLK:=bReset);
+
+// Go back to INIT state on reset
+IF rtReset.Q THEN
+    nState := INIT;
+    IF bError AND stMotionStage.bError AND bCausedError THEN
+        stMotionStage.bReset := TRUE;
+	END_IF
+END_IF
+IF rtExec.Q OR ftExec.Q THEN
+	bDone := FALSE;
+END_IF
+
+CASE nState OF
+	// Start by setting everything to a known value
+    INIT:
+        nState := WAIT_EXEC;
+        bError := FALSE;
```

Comment:
I think we have different ideas of what `bError` means here. I'm taking it to mean "this function block is in an error state", which is absolutely `FALSE` once we've hit `bReset` and are no longer asking for a move. I think you're taking it to mean "the motor associated with this function block is in an error state", which is usually more correct. There's a little bit of nuance here because the request function block error must not be coupled to the motor error if it wasn't making a request.

As implemented is safe because we'll just hop back into error state if we try to ask for a move on an errored axis. I am not expecting people to clear errors without resolving them.

Either way I'm making this comment into a github issue because if our error is coupled to the stage (as it often is) it definitely is more correct to have an exit condition on error state that requires a rising trigger on our `bReset` and a falling trigger on the axis's `bError`.